### PR TITLE
chore: Put secrets straight in env vars, not in shell scripts that will be logged COMPASS-5251

### PR DESCRIPTION
**Pull request by lerouxb**
I also considered writing all the expansions to a yml file and loading that then turning them all into env vars the way mongosh does it somewhere, but that seemed overkill as we can just pass env vars to scripts. Also felt less secure because all the secrets would be in a tmp file. Probably doesn't matter too much.

Yes the branch name doesn't make much sense right now because I realised we can just pass env vars to shell.exec as well so no need to move to subprocess.exec in the same PR.

**Future work:**

* Maybe with the new notary service we won't have to sign through an ssh tunnel on mac, then some hacks can be removed.
* Having serious second thoughts about writing the script that outputs the bash script to eval in javascript, but then my confidence in my bash abilities is pretty low
* We can probably split the env vars so it isn't just one global chunk containing all the secrets. ie. only send the ones necessary for that task. But soon we'll refactor this whole thing anyway and that would be a good time to do it I think.

Here's a green test run: https://spruce.mongodb.com/version/61b092f32a60ed5e66de9a88/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC

To get a sense of the leaked secrets, find a task that ran against the main branch (or anything before this PR), make sure you're on the legacy UI, click "All logs", then Raw. You can search for any of the env vars and you will find them with their values. Compare to the new output: https://evergreen.mongodb.com/task/10gen_compass_main_macos_oneshot_compile_test_package_publish_patch_d3fbd2735dc36fb7db35f0af1729aad9dbd8446d_61b092f32a60ed5e66de9a88_21_12_08_11_12_01#/log/ALL

Try NOTARY_AUTH_TOKEN, NOTARY_SIGNING_KEY, EVERGREEN_AWS_ACCESS_KEY_ID and so on.

### CODE REVIEW

File: .evergreen.yml [hunk 1]
```diff
@@ -46,87 +87,10 @@ functions:
...
-            # To build node modules on RHEL post electron 13 we need
-            # a newer c++ compiler version, this adds it.
-            # https://jira.mongodb.org/browse/COMPASS-5150
-            export PATH="/opt/mongodbtoolchain/v3/bin:\$PATH"
-          fi
-
-          EOF_BUILD_SH
-
-          source ~/compass_env.sh
+          eval $(node scripts/print-compass-env.js);
```
Comment (lerouxb) [inline .evergreen.yml:93]: I could _probably_ try and turn the script into a bash script rather than javascript, but a) I'm pretty terrible at writing shell scripts and b) generating a bash script to be executed using bash felt a bit scary.

My guess is more of this stuff might become javascript as we move it out of evergreen yml anyway.

File: .evergreen.yml [hunk 2]
```diff
@@ -241,32 +198,34 @@ functions:
...
-          export NOTARY_AUTH_TOKEN="${signing_auth_token}"
-          export NOTARY_SIGNING_KEY="${signing_key_name}"
-          export NOTARY_SIGNING_COMMENT="Evergreen project mongodb/compass ${revision} - ${build_variant} - ${branch_name}"
 
           if [ `uname` == Darwin ]; then
             # We do print the environment during the build process in a few places
             # but we probably should stop doing that. After that, we can go back
             # to using plain APPLE_USERNAME and APPLE_PASSWORD environment variables.
-            echo '{"appleUsername": "${apple_username}", "applePassword": "${apple_password}"}' > /tmp/compass-apple-cred.json
+            echo $APPLE_CREDENTIALS > /tmp/compass-apple-cred.json
```
Comment (lerouxb) [inline .evergreen.yml:228]: Must see if we still need this workaround - maybe we can just send the apple credentials as an env var without writing to file first? It wasn't preventing the secret from being leaked in logs anyway because the `echo '{"appleUsername": "${apple_username}", "applePassword": "${apple_password}"}' > /tmp/compass-apple-cred.json` line with expansions expanded gets printed to the logs.

File: .evergreen.yml [hunk 3]
```diff
@@ -14,6 +14,47 @@ variables:
...
+      EVERGREEN_PROJECT: ${project}
+      EVERGREEN_REVISION: ${revision}
+      EVERGREEN_TASK_ID: ${task_id}
+      EVERGREEN_TASK_NAME: ${task_name}
+      EVERGREEN_TASK_URL: https://evergreen.mongodb.com/task/${task_id}
+      EVERGREEN_VERSION_ID: ${version_id}
+      EVERGREEN_WORKDIR: ${workdir}
+      NODE_JS_VERSION: "14.17.5"
+      NPM_VERSION: "7"
+      # secrets
```
Comment (lerouxb) [inline .evergreen.yml:35]: Here I have one pool of compass env vars used by all the shell.exec (or potentially in future subprocess.exec) scripts. We might want to segment it, though? I suppose some only get used by one task.

But we'll probably be refactoring this whole file soon as we split things into smaller tasks, so maybe not worth overthinking it for now.

File: .evergreen.yml [hunk 4]
```diff
@@ -241,32 +198,34 @@ functions:
...
         script: |
           set -e
-          source ~/compass_env.sh
+          eval $(node scripts/print-compass-env.js);
 
+          # TODO: move
           cat <<EOF_BUILD_SH > ~/compass_package.sh
           #! /usr/bin/env bash
+          # TODO: why?
           cd $(pwd)
```
Comment (mcasimir) [inline .evergreen.yml:212]: 🤣 .. yeah as weird as this looks it makes sense, essentially is creating a script that switch immediately to the root of the build no matter where is called from afterwards.. i'm assuming it was necessary cause the directory at that point would have been different 🤷‍♂️

Comment (lerouxb) [reply .evergreen.yml:212]: surely pwd is always the current directory?

Comment (addaleax) [reply .evergreen.yml:212]: @lerouxb Keep in mind that this is something that is being expanded when the script is being written to disk – i.e. `compass_package.sh` itself will contain something like `cd /data/mci/.../src` afterwards instead of a literal `$(pwd)` string

Comment (lerouxb) [reply .evergreen.yml:212]: I'm going to have to deal with that because I'm moving this to an external script.

File: .evergreen/print-compass-env.js [hunk 5]
```diff
@@ -0,0 +1,76 @@
+'use strict';
+
+function maybePrependPath(path, part) {
+  const substring = `${part}:`;
+  if (path.includes(substring)) {
+    return path;
+  }
+  return `${substring}${path}`;
```
Comment (addaleax) [inline .evergreen/print-compass-env.js:8]: This changes the logic a bit – are you sure you want this “maybe” bit here? The order of entries in `PATH` matters a lot, and having duplicates is fine, so it seems like it would be more reliable to always prepend entries?

File: .evergreen/print-compass-env.js [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  # - https://github.com/nodejs/node/issues/34866
+  # - https://github.com/mongodb-js/compass/pull/2403
+  # - https://github.com/mongodb-js/compass/pull/2410
+  */
+  if (pwd.startsWith('\/cygdrive\/c')) {
+    // Change cygdrive from c to z without chanding rest of the path
+    pwd = pwd.replace('\/cygdrive\/c', '\/cygdrive\/z');
+    // we have to change the directory in the shell script we're outputting, not in this node process
+    console.log(`cd ${pwd};`);
+    console.log('echo "Changed cwd on cygwin. Current working dir: $pwd";');
```
Comment (addaleax) [inline .evergreen/print-compass-env.js:39]:
```suggestion
    console.log(`cd "${pwd}"`);
    console.log('echo "Changed cwd on cygwin. Current working dir: $pwd"');
```

File: .evergreen/print-compass-env.js (hunk 5 above)
Comment (lerouxb) [reply .evergreen/print-compass-env.js:8]: I know it matters and that duplicates are fine, but I want this script to be idempotent and only modify the path the first time you run it.

Comment (addaleax) [reply .evergreen/print-compass-env.js:9]: Then maybe you could join all the PATH prepends into a single string, and check whether PATH already starts with that? That way you still fulfill that requirement but still have a solution that gives you a consistent order in the result

Comment (addaleax) [reply .evergreen/print-compass-env.js:9]: Yeah, that’s the point – you check whether `$PATH` starts with `C:B:A:` and prepend `C:B:A:` it if it does not. That way the script is fully idempotent, but you also don’t end up in the situation that `X:Y:A:Z` gets turned into `X:Y:A:Z` but `X:Y:Z` gets turned into `A:X:Y:Z`, where the relative priority of `X` and `A` depends on whether `A` was already in `$PATH` (which might not seem like a big deal, but I would be afraid that that leads to hard-to-debug issues in CI otherwise, where for example `A` might already be in `$PATH` on some machines but not others).

Comment (lerouxb) [reply .evergreen/print-compass-env.js:9]: I'll change it but I think this is all unnecessary because I'm in control of this script that I wrote.

File: .evergreen.yml [hunk 7]
```diff
@@ -241,94 +200,40 @@ functions:
...
-          if [[ "$OSTYPE" == "cygwin" ]]; then
-            echo "Starting Installer Service..."
-            net start MSIServer
-          fi
-
-          # Provide a verbose logging for the release process
-          export npm_config_loglevel="${npm_loglevel}"
-          export DEBUG="${debug}"
+            # write env vars (including secrets) to a script so we can restore them after running ssh
+            .evergreen/export-compass-env.js >> env-vars.sh
```
Comment (lerouxb) [inline .evergreen.yml:216]: Can't really see a better way to restore the environment variables once we ssh for mac. Hopefully we can remove this hack soon once we switch to the new notary service.

File: .evergreen.yml [hunk 8]
```diff
@@ -46,87 +90,11 @@ functions:
...
-            # a newer c++ compiler version, this adds it.
-            # https://jira.mongodb.org/browse/COMPASS-5150
-            export PATH="/opt/mongodbtoolchain/v3/bin:\$PATH"
-          fi
-
-          EOF_BUILD_SH
-
-          source ~/compass_env.sh
+          set -e
+          eval $(.evergreen/print-compass-env.sh)
```
Comment (lerouxb) [inline .evergreen.yml:97]: This is the bit that solves the whole main issue: The expansions that don't need further processing are mixed in as env vars and the ones that do get output as a bash script to be eval'd in here.

The contents of this `script:` block will end up in logs. Previously we did the expansions in the code so it essentially becomes a "hardcoded" block of exports - ie the script block contains the exports with the secrets in plain text. Whereas env vars or the output of .evergreen/print-compass-env.sh don't end up visible in here but only  the few remaining lines of shell script containing no secrets themselves.

File: .evergreen.yml [hunk 9]
```diff
@@ -46,87 +90,11 @@ functions:
...
-            CHANGE_DIR_TO="\$(pwd | sed 's/^\/cygdrive\/c/\/cygdrive\/z/')"
-            cd \$CHANGE_DIR_TO;
-            echo "Changed cwd on cygwin. Current working dir: \$(pwd)"
-          fi
-
-          # Make default evergreen expansions accessible to hadron-build
-          export EVERGREEN_ASSET_PREFIX="s3://mciuploads/${project}/${revision}";
-          export EVERGREEN_AUTHOR="${author}";
-          export EVERGREEN_AWS_ACCESS_KEY_ID="${aws_key}";
-          export EVERGREEN_AWS_SECRET_ACCESS_KEY="${aws_secret}";
```
Comment (lerouxb) [inline .evergreen.yml:75]: Btw I am pretty sure we don't use EVERGREEN_AWS_ACCESS_KEY_ID and EVERGREEN_AWS_SECRET_ACCESS_KEY anywhere so I removed them. If we do use it somewhere I somehow missed it is easy enough to add them back.

**Review by gribnoysup — COMMENTED** [review]
I think we can avoid using this term here

File: .evergreen/print-compass-env.sh [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#! /usr/bin/env bash
+
+export BASHPATH=$PATH
+export OSTYPE=$OSTYPE
```
Comment (mcasimir) [inline .evergreen/print-compass-env.sh:4]: why is this needed?

Comment (lerouxb) [reply .evergreen/print-compass-env.sh:4]: Commented elsewhere but it is a variable only bash sets internally and as far as I can tell it isn't exported. This is how I finally found it https://github.com/dylanaraps/neofetch/issues/433

Comment (mcasimir) [reply .evergreen/print-compass-env.sh:4]: if is available here is available to the subprocess, couldn't you read it from the `.js` file?

Comment (lerouxb) [reply .evergreen/print-compass-env.sh:4]: No and the reason is because it isn't exported. In a terminal type `env | grep OSTYPE` and you'll see it is empty. Then `echo $OSTYPE` and it prints `darwin20.0`

Comment (addaleax) [inline .evergreen/print-compass-env.sh:4]:
```suggestion
export BASHPATH="$PATH"
export OSTYPE="$OSTYPE"
```

**Review by addaleax — APPROVED** [review]
